File: astraquant-backend/services/strategy-service/app/schemas/strategy.py
from datetime import datetime
from pydantic import BaseModel, Field

class CreateStrategyRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    code: str = Field(min_length=1, max_length=128, pattern=r"^[a-zA-Z0-9_\-]+$")
    strategy_type: str = "CONFIG"
    template_id: str | None = None
    description: str | None = None
    tags: list[str] | dict | None = None

class CreateStrategyResponse(BaseModel):
    id: str
    strategy_id: str
    strategy_version_id: str
    name: str
    code: str
    status: str

class UpdateStrategyRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = None
    tags: list[str] | dict | None = None

class CopyStrategyRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    code: str = Field(min_length=1, max_length=128, pattern=r"^[a-zA-Z0-9_\-]+$")

class StrategyListItem(BaseModel):
    id: str
    name: str
    code: str
    strategy_type: str
    status: str
    tags: list[str] | dict | None = None
    latest_version: str | None = None
    latest_version_id: str | None = None
    created_at: datetime | None = None

class StrategyVersionSummary(BaseModel):
    id: str
    version: str
    status: str
    params_hash: str
    created_at: datetime | None = None

class StrategyDetail(BaseModel):
    id: str
    name: str
    code: str
    strategy_type: str
    status: str
    description: str | None = None
    tags: list[str] | dict | None = None
    latest_version_id: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    versions: list[StrategyVersionSummary]

class StatusChangeRequest(BaseModel):
    reason: str = Field(min_length=1)

class StatusChangeResponse(BaseModel):
    strategy_id: str
    status: str

File: frontend/src/types/strategy.ts
export interface StrategyTemplate {
  id: string;
  code: string;
  name: string;
  strategy_type: string;
  description?: string | null;
  default_config?: Record<string, unknown> | null;
}

export interface StrategyTemplateResponse {
  items: StrategyTemplate[];
  total?: number;
}

export interface StrategyListItem {
  id: string;
  name: string;
  code: string;
  strategy_type: string;
  status: string;
  tags?: string[] | Record<string, unknown> | null;
  latest_version?: string | null;
  latest_version_id?: string | null;
  created_at?: string | null;
}

export interface StrategyListResponse {
  items: StrategyListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface StrategyVersionSummary {
  id: string;
  strategy_id?: string;
  version: string;
  status: string;
  params_hash: string;
  created_at?: string | null;
}

export interface StrategyVersionDetail extends StrategyVersionSummary {
  strategy_id: string;
  params_json: Record<string, unknown>;
  risk_params_json: Record<string, unknown>;
  source_version_id?: string | null;
}

export interface StrategyVersionListResponse {
  items: StrategyVersionSummary[];
  total: number;
}

export interface CreateStrategyVersionPayload {
  source_version_id?: string;
  change_reason: string;
  params_json?: Record<string, unknown>;
  risk_params_json?: Record<string, unknown>;
}

export interface UpdateStrategyVersionParamsPayload {
  params_json: Record<string, unknown>;
  risk_params_json: Record<string, unknown>;
}

export interface StrategyDetail extends StrategyListItem {
  description?: string | null;
  updated_at?: string | null;
  versions: StrategyVersionSummary[];
}

export interface CreateStrategyPayload {
  name: string;
  code: string;
  description?: string;
  strategy_type: string;
  template_id?: string;
  tags?: string[];
}

export interface UpdateStrategyPayload {
  name?: string;
  description?: string;
  tags?: string[];
}

export interface CopyStrategyPayload {
  name: string;
  code: string;
}
